mono/mini/llvm-jit.cpp

```diff
@@ -247,24 +249,82 @@ init_mono_llvm_jit ()
 }
 
 static MonoLLVMJIT *
-make_mono_llvm_jit (TargetMachine *target_machine)
+make_mono_llvm_jit (TargetMachine *target_machine, LLVMModuleProviderRef mp)
 {
 	return new MonoLLVMJIT{target_machine};
 }
 
 #elif LLVM_API_VERSION > 600
 
+// The OptimizationList is automatically populated with registered Passes by the
+// PassNameParser.
+//
+static cl::list<const PassInfo*, bool, PassNameParser>
+PassList(cl::desc("Optimizations available:"));
+
 class MonoLLVMJIT {
 public:
 	/* We use our own trampoline infrastructure instead of the Orc one */
 	typedef RTDyldObjectLinkingLayer ObjLayerT;
 	typedef IRCompileLayer<ObjLayerT, SimpleCompiler> CompileLayerT;
 	typedef CompileLayerT::ModuleHandleT ModuleHandleT;
 
-	MonoLLVMJIT (TargetMachine *TM, MonoJitMemoryManager *mm)
+	MonoLLVMJIT (TargetMachine *TM, MonoJitMemoryManager *mm, LLVMModuleProviderRef mp)
 		: TM(TM), ObjectLayer([=] { return std::shared_ptr<RuntimeDyld::MemoryManager> (mm); }),
 		  CompileLayer (ObjectLayer, SimpleCompiler (*TM)),
-		  modules() {
+		  modules(),
+		  fpm (unwrap (mp)) {
+		initPassManager ();
+	}
+
+	void initPassManager () {
+		// Initialize passes (copied from https://github.com/mono/llvm/blob/release_60/tools/opt/opt.cpp#L379)
+		PassRegistry &registry = *PassRegistry::getPassRegistry();
+		initializeCore(registry);
+		//initializeCoroutines(registry);
+		initializeScalarOpts(registry);
+		//initializeObjCARCOpts(registry);
+		//initializeVectorization(registry);
+		//initializeIPO(registry);
+		initializeAnalysis(registry);
+		initializeTransformUtils(registry);
+		initializeInstCombine(registry);
+		//initializeInstrumentation(registry);
+		initializeTarget(registry);
+		//initializeExpandMemCmpPassPass(registry);
+		//initializeScalarizeMaskedMemIntrinPass(registry);
+		//initializeCodeGenPreparePass(registry);
+		//initializeAtomicExpandPass(registry);
+		//initializeRewriteSymbolsLegacyPassPass(registry);
+		//initializeWinEHPreparePass(registry);
+		//initializeDwarfEHPreparePass(registry);
+		//initializeSafeStackLegacyPassPass(registry);
+		//initializeSjLjEHPreparePass(registry);
+		//initializePreISelIntrinsicLoweringLegacyPassPass(registry);
+		//initializeGlobalMergePass(registry);
+		//initializeIndirectBrExpandPassPass(registry);
+		//initializeInterleavedAccessPass(registry);
+		//initializeEntryExitInstrumenterPass(registry);
+		//initializePostInlineEntryExitInstrumenterPass(registry);
+		//initializeUnreachableBlockElimLegacyPassPass(registry);
+		//initializeExpandReductionsPass(registry);
+		//initializeWriteBitcodePassPass(registry);
+
+		// FIXME: find optimal mono specific order of passes, see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
+		const char *opts = "opt -simplifycfg -sroa -instcombine -early-cse-memssa -instcombine -gvn";
+		char **args = g_strsplit (opts, " ", -1);
```

Comment:
What is "opt ?

Reply:
it ignores the first argument for some reason so it's fake

Reply:
Actually a leading space helps